### ISSUE TITLE
Blogging assistance: Things Rust doesn't let you do

### ISSUE THREAD

**Issue by golddranks** [rust-community/content-o-tron#61]
Hi! I've been writing a long blog post / survey about the current limitations of the borrow checker. I try to explain why the limitations apply and review some upcoming features that are going to lift them.

I need help with grammar (no need to bother with heavy copy editing, just check if there's some obviously awkward word choices or grammar) plus a technical review.

Here's a link to the draft:
https://medium.com/@GolDDranks/things-rust-doesnt-let-you-do-draft-f596a3c740a5

**Comment by booyaa** [comment]
Sorry for the delay, we'll allocate a reviewer.

**Comment by golddranks** [comment]
Any news about this?

**Comment by Amanieu** [comment]
> but there is still some cases

s/is/are/

> Because there is not shared ownership built into the language

s/not/no/

> References are thus, always valid.

You might want to add a sentence about references always being non-null as well.

> Rust doesn’t only check that you don’t mutate the value “through” a reference if you aren’t allowed to, it ensures that there exists no other references that could mutate it. Even the owner isn’t allowed to mutate the value while there exists some references to it — the value is effectively frozen.

These sentences are unclear, and doesn't clearly explain 1 writer XOR multiple readers. Maybe this wording is better: As long as a mutable reference to a value exists, that value can only be accessed (read or written) through that reference. All other references —  including the owner —  are effectively frozen while the mutable reference exists.

> There is some good reasons

s/is/are/

> That means that if you pass a value somewhere, you lose the ownership over it and can’t use it anymore. 

Maybe mention `Copy` types as an exception to this rule?

> Other than that, there is some additional reasons: one of them has to do with processor caches and another has to do with thread safety. To quickly explain the gist: single mutability ensures that you can’t invalidate cached versions of pointed values by mutating that value,

The first reason (aliasing) actually has nothing to do with processor caches, and is all about compiler optimization: when a piece of code holds an immutable reference the compiler can safely assume that the data pointed to will not be modified as long as the immutable reference exists, even after passing that reference to a function which the compiler doesn't know about. This is unlike C/C++ where that function could remove `const` from a pointer and modify the value (this is allowed by the C standard, so the compiler must be more conservative).

> there isn’t any other mutable references

s/isn't/aren't/

> I’d like to review some cases where it would be certainly safe to do things the borrow checker won’t allow, but for one reason or another, isn’t sophisticated enough to see that.

Try this instead: I’d like to review some cases that are certainly safe, but for one reason or another, the borrow checker isn’t sophisticated enough to see that.

> This is a very simple way to think of the borrows — and a simplistic one.

You're repeating yourself (simple x2).

> Another example is iterators: you can mutably iterate over a vector, and as a result get a mutable reference to each of the elements.

That doesn't really solve the problem though since you're still only limited to a single mutable reference to an element at a time.

> There isn’t any proposals

s/isn't/aren't/

> I’m imagining a facades

s/facades/facade/

> It must be more careful with cache invalidation

s/cache invalidation/aliasing/

> [A general note regarding the section on `Cell`]

You do not mention anywhere that `Cell` only works on `Copy` types. Non-`Copy` types need to use `RefCell` instead, which has some runtime overhead (and obviously you can't convert `&mut T` to `&RefCell<T>` since there is extra runtime metadata).

> there is basically two kinds of data types

s/is/are/

> there are subtler reasons around caches and threading too

s/caches/aliasing/

> `list.retain(|i| filter.allowed(i));`

Should be `self.list.retain`

> there is going to problems

there *are* going to *be* problems

> There has been some ideas

s/has/have/

> It would be sound in principle to [...] It would be sound in principle to [...]

Avoid using the same wording in two consecutive sentences, it's distracting.

> However, in the future I’d like Rust to go all the way

You already started the previous paragraph with "However". Just remove it here.

**Comment by Amanieu** [comment]
Great article! :+1:

**Comment by golddranks** [comment]
Great! Thank you!

**Comment by golddranks** [comment]
I'm going through the corrections now! A question: I thought that simple and simplistic mean different things. (Simple → in a good way, simplistic → made too simple, missing sophistication) Is there any other clearer words for that?

Re: processor cache invalidation. Indeed, my bad! I was thinking of "caching values in memory in registers without re-loading them from memory because of aliasing restrictions", but that's entirely different from actual processor caches.

I'll be asking more clarifications in a bit!

**Comment by golddranks** [comment]
> That doesn't really solve the problem though since you're still only limited to a single mutable reference to an element at a time.

The Iterator trait allows each mutable reference to live as long as the iterator itself, which means that you can, for example, collect the mutable references to a collection where they live simultaneously.

> You do not mention anywhere that Cell only works on Copy types. Non-Copy types need to use RefCell instead, which has some runtime overhead (and obviously you can't convert &mut T to &RefCell<T> since there is extra runtime metadata).

Cell works with non-Copy types too nowadays. ( https://github.com/rust-lang/rfcs/blob/master/text/1651-movecell.md )

Overall, I seem to have a problem with "there is/are" patterns :P. Fixed those; thank you!

**Comment by Amanieu** [comment]
> I'm going through the comments now! A question: I thought that simple and simplistic mean different things. (Simple → in a good way, simplistic → made too simple, missing sophistication) Is there any other clearer words for that?

While it is true that simple and simplistic have different meanings, using both in the same sentence still sounds bad to someone reading it. How about this wording instead: "This is a very simple way to think of the borrows —but a rather unsophisticated one."

> The Iterator trait allows each mutable reference to live as long as the iterator itself, which means that you can, for example, collect the mutable references to a collection where they live simultaneously.

That won't work since `Iterator::next` takes `&mut self`. This prevents you from calling `next` as long as there exists a reference (of any type) derived from the lifetime of the iterator (the iterator is already mutably borrowed).

**Comment by golddranks** [comment]
I got busy with other things in life, and I forgot to publish this. Now I came back to it. Published it on Reddit and HN. Thanks a bunch for the review!!!
